File: lecture17/bst.cpp
// bst: binary search tree

#include <iomanip>
#include <iostream>

using namespace std;

// Structures

typedef struct Node Node;
struct Node {
    int   value;
    Node *left;
    Node *right;
};

typedef struct Tree Tree;
struct Tree {
    Node *root;
};

// Functions

Node*	tree_insert_r(Tree *tree, Node *node, int value) {
    // Base case: No node, so create one
    if (node == nullptr) {
    	return new Node{value, nullptr, nullptr};
    }

    // Recursive: go left if value is than current node value
    if (value < node->value) {
    	node->left  = tree_insert_r(tree, node->left, value);
    } else {   // go right
    	node->right = tree_insert_r(tree, node->right, value);
    }

    return node;
}

void	tree_insert(Tree *tree, int value) {
    tree->root = tree_insert_r(tree, tree->root, value);
}

bool	tree_search_r(Tree *tree, Node *node, int value) {
    // Base case: No node, so missing
    if (node == nullptr) {
    	return false;
    }

    // Base case: Found
    if (value == node->value) {
    	return true;
    }

    // Recursive: go left if value is than current node value
    if (value < node->value) {
    	return tree_search_r(tree, node->left, value);
    } else {   // go right
    	return tree_search_r(tree, node->right, value);
    }
}

bool	tree_search(Tree *tree, int value) {
    return tree_search_r(tree, tree->root, value);
}

void	tree_print_r(Tree *tree, Node *node) {
    if (node == nullptr) {
    	return;
    }

    tree_print_r(tree, node->left);
    cout << node->value << endl;
    tree_print_r(tree, node->right);
}

void	tree_print(Tree *tree) {
    tree_print_r(tree, tree->root);
}

// Main Execution

int main(int argc, char *argv[]) {
    Tree *tree = new Tree();

    for (int i = 1; i < argc; i++) {
    	tree_insert(tree, atoi(argv[i]));
    }
    
    for (int i = 1; i < argc; i++) {
    	int number = atoi(argv[i]);
    	cout << number   << " " << boolalpha << tree_search(tree, number) << endl;
    	cout << 2*number << " " << boolalpha << tree_search(tree, 2*number) << endl;
    }

    tree_print(tree);

    return 0;
}

File: lecture17/bst.py
#!/usr/bin/env python3

from dataclasses import dataclass
import sys

# Classes

'''
class Node(object):
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left  = left
        self.right = right

    def __str__(self):
        return f'Node({self.value},{self.left},{self.right})'
'''

@dataclass
class Node:
    value:  int
    left:   'Node' = None
    right:  'Node' = None

class Tree:
    def __init__(self):
        self.root = None

    def search(self, value):                # Discuss: Search
        return self._search(self.root, value)

    def _search(self, node, value):        # Discuss: Recursive helper
        if node is None:
            return False

        if value == node.value:
            return True

        if value <= node.value:
            return self._search(node.left, value)
        else:
            return self._search(node.right, value)

    def insert(self, value):                # Discuss: Insertion
        self.root = self._insert(self.root, value)

    def _insert(self, node, value):
        if node is None:
            return Node(value)

        if value <= node.value:
            node.left  = self._insert(node.left, value)
        else:
            node.right = self._insert(node.right, value)

        return node

    def print(self):                        # Discuss: Traversal
        self._print(self.root)

    def _print(self, node):                # Discuss: In-Order Traversal
        if not node:
            return

        self._print(node.left)
        print(node.value)
        self._print(node.right)

    def nodes(self):                        # Review: Generators
        yield from self._nodes(self.root)

    def _nodes(self, node):
        if node:
            yield from self._nodes(node.left)
            yield node.value
            yield from self._nodes(node.right)

# Main Execution

def main():
    tree = Tree()
    for number in map(int, sys.argv[1:]):
        tree.insert(number)

    for number in map(int, sys.argv[1:]):
        print(number, tree.search(number))
        print(2*number, tree.search(number*2))

    tree.print()
    print(', '.join(map(str, tree.nodes())))

if __name__ == '__main__':
    main()
